### DISCUSSION TITLE
applying a feature patch when its name has changed

### DISCUSSION THREAD

**Discussion by gireeshpunathil** [eclipse-pde/eclipse.pde#1288] [Q&A]
I have a feature of which the name has changed in a later version. I created a feature patch with the new name, but the p2 installer is refusing to install the patch, as it doesn't know the parent feature on which to install this. any ideas?

this is the error I get: `Equinox P2 CORE Feature Patch (bugzillas 128,247)” is not applicable to the current configuration and will not be installed.`

thanks in advance!

**Comment by merks** [comment]
It sounds like you are trying to install this when the installation still has the old name.  Is that the case?  
- If so, that doesn't really make sense because you can't patch something that isn't installed.
- If not, then I don't see how it having an older name is even relevant.

**Comment by gireeshpunathil** [comment]
@merks - thanks. yes, I am trying to install this patch (with new name, `bcpg`) when the installation still has the old name (`org.bouncycastle`).

I understand the installer cannot figure out the parent feature for which this is a patch, but this is where we are! any workaround is appreciated!

**Reply by merks** [reply]
I’m more confused though because this is a bundle not a feature so I have no mental model for what exactly you are doing.

**Comment by gireeshpunathil** [comment]
oops! apologies for confusing you. a feature that consumed `org.bouncycastle` earlier, now wants to be patched which consume `bcpg` in lieu of of the former - that is the scenario which resulted in this error.

**Reply by merks** [reply]
Are you sure the error is related to the bouncycastle thing at all?  Note the we’ve generally been removing includes of 3rd party plugins.

**Comment by gireeshpunathil** [comment]
https://github.com/eclipse-equinox/p2/blob/R4_23_maintenance/features/org.eclipse.equinox.p2.core.feature/feature.xml#L139-L141

yes, I guess this dependency is removed from master, (may be this: https://github.com/eclipse-equinox/p2/commit/f885ca80b24eae6e4a6b84b10552eb3e96b73d12) but we do have it still in 4.23 maintenance. I could remove from there too, but unsure about any implications! also would like to know the recommended method of patching in such scenarios (in cases where we really need those).

thanks!

**Reply by laeubi** [reply]
Main implication is that maybe both bundles will be installed, but feature patches are somehow "special" ...

**Reply by merks** [reply]
Yes, if you don't specify it in a feature, nothing is likely to update the bundle itself.  Moreover nothing is likely to remove the older named bundle and replace it with the newer named bundle.  But I still don't understand why a feature patch should not allow an additional bundle to be added or some bundle to be removed from the feature itself.  You might even need a negative requirements via p2.inf to remove the older named versions from the installation.  It's definitely a trick problem!

**Reply by merks** [reply]
Something that might work is to use p2.inf to specify an IU with the old name, but a newer version, specify that in the feature patch, and synthesize that IU to require the newly named IU so that the update would be happy to see the old thing being updated and that update would also force the new thing to be installed:

https://wiki.eclipse.org/Equinox/p2/Customizing_Metadata

**Reply by laeubi** [reply]
Every Iu has an Update ID so one *might* be able to make it even updatable, but in the end why bother at all? P2 will simply fetch the new one if it is required, the worst that can happen is that the old lingers around but no one is using that...

While with enforce it to uninstall, it might lead to the patch being not installable at all because someone else in the system still need it...
